jina/executors/indexers/query/keyvalue.py

```diff
@@ -0,0 +1,58 @@
+from typing import Iterable
+
+from jina.executors.dump import import_metas
+from jina.executors.indexers.keyvalue import BinaryPbIndexer
+from jina.executors.indexers.query import BaseQueryIndexer
+
+
+class QueryBinaryPbIndexer(BinaryPbIndexer, BaseQueryIndexer):
+    """A write-once Key-value indexer."""
+
+    def __init__(self, dump_path=None, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if dump_path:
+            self.load_dump(dump_path)
+        else:
+            self.logger.error(
+                f'Dump path for {self.__class__} was None. No data to load for {self.__class__}'
+            )
+
+    def load_dump(self, path):
+        """Load the dump at the path
+
+        :param path: the path of the dump"""
+        ids, metas = import_metas(path, str(self.pea_id))
+        self.add(list(ids), list(metas))
+        self.write_handler.flush()
+        self.write_handler.close()
+        self.handler_mutex = False
+        self.is_handler_loaded = False
+        del self.write_handler
+        # warming up
+        self.query('someid')
+        # the indexer is write-once
+        self.add = lambda *args, **kwargs: self.logger.warning(
+            f'Index {self.index_abspath} is write-once'
+        )
```

Comment:
I think the code is cleaner, if you call `super.add(...)` instead of `self.add` in line 25. Then you can overwrite the `add` method in the same way as `update` and `delete`. Some sample code showcasing the behavior:
```
class A:
    def add(self):
        print("a")
        self.add()


class B(A):
    def add(self):
        print("b")

    def call_super(self):
        super().add()


B().call_super()
> a
> b
```

Reply:
This is dangerous, what if u have an instance of `A`?

Reply:
Just did a lot of refactoring there. Extracted the writing into a BinaryPbMixin and then subclassed it for both the DBMS and the original Indexers. Can you check again?

Reply:
> This is dangerous, what if u have an instance of `A`?

The recursive call of `self.add` was just for demonstration purpose. For sure, you shouldn't do that.

Reply:
Now you definitely can handle the `add` the same way as `update` and `delete, since we only call `_add` above.